1-tolean.py

```diff
@@ -493,12 +496,40 @@ def check_opt(opt, timeout, bitwidth, hide_progress):
 
   return True # succeeded, did not time out
 
+def sanitize_name(name):
+  renamed = re.sub(r'[()~&>|^=]', '', name)
+  renamed = re.sub(r'[:, -]', '_', renamed)
+  return renamed
+
+def gather_names_widths(constants):
+  names_widths = {}
+  for name in constants:
+    bw = unify_bitwidths([x.bitwidth for x in constants[name]])
+    if bw not in names_widths:
+      names_widths[bw] = [name]
+    else:
+      names_widths[bw].append(name)
+  return names_widths
 
 def print_as_lean(opt):
   name, pre, src, tgt, ident_src, ident_tgt, used_src, used_tgt, skip_tgt = opt
-  (src_str, src_state) = to_lean_prog(src, num_indent=2, skip=[])
-  (tgt_str, tgt_state) = to_lean_prog(tgt, num_indent=2, skip=[])
-
+  print("dbg> printing " + name + " as lean")
+  (src_str, src_state, src_bw) = to_lean_prog(src, num_indent=2, skip=[])
+  (tgt_str, tgt_state, tgt_bw) = to_lean_prog(tgt, num_indent=2, skip=[], expected_bitwidth=src_bw, constants=src_state.constant_names)
+  bitwidth = unify_bitwidths([src_bw, tgt_bw])
+  constant_decls = ""
+  names_widths = gather_names_widths(tgt_state.constant_names)
+  for w in names_widths.iterkeys():
+    constant_decls += "("
+    constant_decls += " ".join([x  for x in names_widths[w]])
+    constant_decls += " : Bitvec " + str(w) + ")\n"
+  for w in tgt_state.constant_names.iterkeys():
+    assert w in src_state.constant_names
+  print("dbg> lhs bw: " + str(src_bw) + " rhs bw: " + str(tgt_bw) + " unified to: " + str(bitwidth))
+  if bitwidth == 'w' or src_str.find(" w ") != -1 or tgt_str.find(" w ") != -1:
+    forall_stmt = " : forall (w : Nat) "
+  else:
+    forall_stmt =  ": forall "
```

Comment:
Hm, I had replied to this, I wonder what happened 🤔 

@bollu I would prefer not to, as it is very inconsistent with the rest of the code: other variables you define there have more lines between their definition and usage, like `name` or `tgt_str`, why does this comment not apply to those? Moreover, I agrue it's less readable if it breaks the chain of `out +=`s, this way the flow is very clear: define all the strings we need and then build the output, it would be very weird to instersperse a definition there in between.